### PR TITLE
Remote copy/paste

### LINKED ISSUE #914: Copy/Paste to remote instance
There are ongoing attempts from [past issues](https://github.com/neovide/neovide/issues?q=copy+paste) about copy/pasting. Most of them are rejected because the author doesn't consider copy/pasting should be handled by GUI applications.

In normal scenarios, the GUI doesn't need to handle copy pasting as it works internally, and handled by neovim. However, when using GUI remotely (TCP or WSL), the GUI must handle copy pasting.

It is the same concept as ssh terminals, you copy text from the local machine, and paste the text to the remote machine. This is what Neovide is lacking of. I often find myself remote to a server, and can't directly copy the text/solution I've found on the local machine. Which is tedious because I would have to type the text manually. And in cases where the text is long, or a checksum is to be entered, it would be inconvenient.

Consider the scenario "Configuring nginx"
- `ssh a.b.c.d`; `cd /etc/nginx`; `nvim --headless --listen 0.0.0.0:6666 &`; `disown`, logout
- `neovide --disowned --remote-tcp=a.b.c.d:6666`
- Go to https://www.digitalocean.com/community/tools/nginx?global.app.lang, setup and copy the text
- In Neovide, paste the solution (can't because Neovide doesn't send the clipboard to remote instance)

Neovim-qt has a discussion about this, and a PR
https://github.com/equalsraf/neovim-qt/pull/479
https://github.com/equalsraf/neovim-qt/issues/298

To make it short, Neovide should provide a method to write text from the local system clipboard to the remote instance, whether it's over TCP or WSL.

We can chose to implement as method 1 or 2.
1. Override `"+p` when using remote mode, completely ignore the remote system clipboard (modify `+` register)
2. Create a command `:NeovidePaste` to write the current system clipboard at cursor

### PR DESCRIPTION
**Pull request by nganhkhoa**
Closes #914 

- [x] Setup rpcrequest
- [x] Set `g:clipboard` when remote
- [x] Copy handler `+` register
- [x] Paste handler `+` register
- [x] Copy handler `*` register
- [x] Paste handler `*` register
- [ ] Toggle command
- [x] Variable to override on init

I decided to hook the `*` register to same remote clipboard (A quick solution). A variable `g:neovide_no_custom_clipboard` is introduced and disable custom clipboard hook.

Tests to be done:
- [x] Remote headless
- [ ] WSL

Though it should work with WSL, but I don't have WSL configured. SSH instances, and `same host, same client` are tested.

Edge cases:
- [x] Empty clipboard
- [x] Clipboard is not text (picture)
- [ ] Copy race condition (hardly, but should handle) (1)

Empty clipboard and clipboard not text will returns Error, and neovim outputs `Nothing in register +`.

Haven't done anything related to race condition, I assume it has problem, however never checked. Current PR is suffice.

Note:
1. Because copy issue `rpcnotify`, we need to move into Parallel/Serial command to handle race condition. Race only when record/replay or manually copy many times in a small time frame. Paste don't have this problem, since paste is `rpcrequest`, which neovim waits until the request returns.

### CODE REVIEW

File: src/bridge/handler.rs [hunk 1]
```diff
@@ -36,6 +39,34 @@ impl NeovimHandler {
 impl Handler for NeovimHandler {
     type Writer = TxWrapper;
 
+    async fn handle_request(
```
[the remark this answers is not in the document]
Comment (nganhkhoa) [reply src/bridge/handler.rs:42]: I also plan to do that, this was just a poc that prove the code works.

File: src/bridge/mod.rs [hunk 2]
```diff
@@ -268,6 +268,34 @@ async fn start_neovim_runtime(
         .await
         .ok();
 
+    if let ConnectionMode::RemoteTcp(_) = connection_mode() {
+        // this should be default, unless disabled with `--no-clipboard`
+        // we also provide a separated setup command and disable command
+        let custom_clipboard = r#"
```
[the remark this answers is not in the document]
Comment (nganhkhoa) [reply src/bridge/mod.rs:274]: If we are strict on code organization, having a separate code serve as the vim init for neovide is a good goal. But we should weight in performance between rpc call and script load.

Another point, as neovide ship as a standalone binary, the script is embedded. We can write the script into autoload folder on first run, and check if `g:neovide` is loaded, if `g:neovide_script_version` is lower, we update the script. Just my thoughts on this. Well be a new issue for this feature.

Placing scripts in `~/.local/share/nvim/site/autoload` or `~/.config/nvim/autoload` can make neovim loads the script on runtime, similar to how package manager works.

**Comment by nganhkhoa** [comment]
After a few months with this, I think the big issue this has is probably how to re-register the command with the correct channel id. As I've been experiencing #1010 for so many times. When I restarted neovide, the copy/paste channel is wrong.

In the next few days, I'll rebase and update the other todos.

**Comment by Kethku** [comment]
@nganhkhoa I think #1010 is *FINALLY* fixed. If you're still interested try giving a new build a go

**Comment by nganhkhoa** [comment]
Rebased, pushed commits, seperate files for clipboard.

I will test the edge cases soon. Been using this privately for few months, it should work.

File: src/bridge/clipboard.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+use clipboard::ClipboardContext;
+use clipboard::ClipboardProvider;
+
+pub fn get_remote_clipboard() -> Result<Value, Box<dyn Error>> {
+    let mut ctx: ClipboardContext = ClipboardProvider::new()?;
+    let lines = ctx
+        .get_contents()?
+        .replace("\r", "")
+        .split("\n")
```
Comment (fredizzimo) [inline src/bridge/clipboard.rs:13]: I think this line ending conversion should be configurable, since it depends on both the source and host OS, but there could be some autodetected defaults. 

Also currently it looks like it converts from Windows endings to Linux endings, when pasting pasting into Neovim, but when copying from Neovim, it does not not do the opposite. That is different from how the default `win32yank.exe` implementation works on WSL, which converts both ways https://github.com/neovim/neovim/blob/22b538139606321bb1083665c510e81adb3d6670/runtime/autoload/provider/clipboard.vim#L130

Comment (nganhkhoa) [reply src/bridge/clipboard.rs:13]: Fair points. Should we look for current file endline encoding and make it compatible with that?

get clipboard: check current endline encoding, convert?
set clipboard: no furthur changes, direct code copy

Comment (fredizzimo) [reply src/bridge/clipboard.rs:13]: I'm not sure to be honest. 

I'm note sure, but I think nvim always internally uses `LF`, so stripping the `CR` like you do here should be fine for pasting, otherwise I guess the current file encoding is a better option. 

For copying you probably want to use the convention of the operating system in which Neovide is running, so `CR LF` on Windows.

But ideally it should be configurable, and even configurable at runtime. There has been a few instances where I have wished that I could copy in the Linux format. So for people that needs that in a common scenario could even make a special keybinding. But I think it's fine to just use a config variable for this, since you can map a custom function that sets the variable, copies  and then resets it.

Comment (Kethku) [reply src/bridge/clipboard.rs:13]: I think this is the way to go for now. If in the future we find a way to make it work automatically, no reason we can't fix it later. One bit though: the default for the setting should be set by platform

File: src/bridge/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kethku) [reply src/bridge/mod.rs:267]: Very delayed response 😅...

I like this thinking except for one issue, writing into the local machine's autoload folder would mean that those same scripts would get loaded every time afterward even if run in terminal or with another gui. This would likely have a performance burden on other systems.

I'm still thinking about this, but for now the embedded approach of manually running nvim commands on startup is sufficient.

[the remark this answers is not in the document]
Comment (fredizzimo) [reply src/bridge/mod.rs:267]: Another comment regarding this. We also need to be careful to load this before `init.vim`, so that users can override `g:clipboard` if they want.

**Comment by nganhkhoa** [comment]
Hmm, I want to ask why can't I log or print anything in bridge?

I only have `Neovim notification: "redraw"` printed. Is the logger needs any extra config to be usable?

`println!`, `log.info!`, `log.trace!` in `bridge/handler` and `bridge/clipboard` have no effect. Tried debug and release build as well.

**Comment by nganhkhoa** [comment]
Notice a big code change in main, rebased, the print log works now.

**Comment by nganhkhoa** [comment]
I updated the PR. Compatible with the lastest file structure changes.

On copy: line endings are chained with remote system line endings.
On paste: line endings matches the current file format (check with `set ff`).

**Comment by nganhkhoa** [comment]
I also tested with

- multiple instances: open 2+ neovide remote instances.
- disconnect and reconnect: open neovide remote instance, close (click x button), open another remote instance.

**Comment by Kethku** [comment]
Amazing. I will try this when I get home and merge

**Comment by Kethku** [comment]
Tested in wsl. Works perfectly. Great work, thanks a bunch

**Comment by fredizzimo** [comment]
@Kethku, just a question to be sure. As far as I can see this does nothing on a standard WSL configuration, so I assume that you tested with a TCP connection to WSL? And without Win32Yank.exe in the path, just to make sure that it's not used?

**Comment by Kethku** [comment]
did you set clipboard to unnamedplus?

Let me see what happens without the change. Maybe clipboard just worked transparently and I didn't notice...

**Comment by fredizzimo** [comment]
I did not test it myself, but 
https://github.com/neovide/neovide/blob/ab0394ae40a466c5e4c7c53fd4ece85fc52f0c24/src/bridge/setup.rs#L144-L146

Should only affect TCP connections, that's why I'm asking if you tested it properly, and not just the old WSL behaviour with Win32Yank.

**Comment by Kethku** [comment]
you're likely right. Looks like Win32Yank was indeed on my path. I think this probably should be changed to include wsl. I will make the change and ensure that I'm testing without the utility on the path

**Comment by Kethku** [comment]
Thanks for the spot check :) This is what I get for multi tasking

**Comment by fredizzimo** [comment]
WSL support would be nice at least to enable a good out of the box experience for new users. But I think most users still want to use Win32Yank, since that works in the terminal too. Yes, sometimes I open nvim from the terminal... 

Personally I need an even more complex setup, since I have WSLg and xclip installed, and neovim would therefore default to xclip, so I had to use this 
```
let g:clipboard = {
          \   'name': 'win32yank-wsl',
          \   'copy': {
          \      '+': 'win32yank.exe -i --crlf',
          \      '*': 'win32yank.exe -i --crlf',
          \    },
          \   'paste': {
          \      '+': 'win32yank.exe -o --lf',
          \      '*': 'win32yank.exe -o --lf',
          \   },
          \   'cache_enabled': 0,
          \ }
```
I don't remember what issues xclip caused, since I think that it should be synchronized with the windows clipboard on WSLg, but it was probably some line ending issue. It might also have been that the synchronization did not work properly. I don't really use WSLg that much, since it's so buggy so I don't know.

**Comment by Kethku** [comment]
I'm wondering if Win32Yank is truly better than the gui version? Maybe a world where you can override the g:clipboard if you want, but it defaults to using the remote system is ok?

**Comment by fredizzimo** [comment]
Well, currently with these changes you can set `g:neovide_no_custom_clipboard`, and then I think the default builtin `g:cllpboard` should be used. And you should also be able to override it in your init file. 

This should probably cover all cases already, even if we add WSL support. If you don't run through Neovide, then the custom function is not installed, and it would revert back to default. If you run through Neovide, then the custom version is used, but unless it's broken, I don't see much reason to disable it. And if you really want, you can.

**Comment by nganhkhoa** [comment]
How does neovide connect with neovim running in WSL instance? Just need to set `is_remote=true` to WSL session.
